filebeat/docs/modules-dev-guide.asciidoc

```diff
@@ -0,0 +1,304 @@
+Beats https://github.com/elastic/beats/blob/master/CONTRIBUTING.md[CONTRIBUTING]
+guide.
+
+=== Overview
+
+Each Filebeat module is composed by one or more "filesets". We usually create a
+module for each service that we support (`nginx` for Nginx, `mysql` for Mysql,
+etc.) and a fileset for each type of logs that that service creates. For
+example, the Nginx module has `access` and `error` filesets. You can contribute
+a new module (with at least one fileset) or only a new fileset into an existing
+module.
+
+=== Creating a new fileset
+
+Regardless of whether you are creating a fielset in a new module or in an
+existing module, the procedure is similar. Run the following command in the
+`filebeat` folder:
+
+[source,bash]
+----
+make create-metricset
+----
+
+You'll be prompted to enter a module and metricset name. Only use characters
+`[a-z]` and, if required, underscores (`_`). No other characters are allowed.
+For the module name, you can us either a new module name or an existing module
+name. If the module doesn't exist, it will be created.
+
+In this guide we use `{fileset}` and `{module}` as placeholders for the fileset
+and module names. You need to replace these with the actual names you entered in
+the previous command.
+
+After running the above make command, you'll find the fileset, along with its
+generated files, under `module/{module}/{fileset}`. This directory
+contains the following files:
+
+[source,bash]
+----
+module/{module}/{fileset}
+├── manifest.yml
+├── config
+│   └── {fileset}.yml
+├── ingest
+│   └── pipeline.json
+├── _meta
+│   └── fields.yml
+└── test
+----
+
+Let's look at these files one by one.
+
+==== manifest.yml
+
+The `manifest.yml` is the control file for the module, where variables are
+defined and the other files are referenced. It is a YAML file, but in many
+places in it, built-in or defined variables can be used using the
+`{{.variable}}` syntax.
+
+The `var` section of the file defines the fileset variables and their default
+values. The module variables can be referenced in the rest of the configuration
+files, and their value can be overridden at runtime via the Filebeat
+configuration.
+
+As the fileset creator, you can use any names for the variables you define. Each
+variable must have a default value. So in it's simplest form, this is how you
+can define a new variable:
+
+[source,yaml]
+----
+var:
+  - name: pipeline
+    default: with_plugins
+----
+
+Most fileset should have a `paths` variable defined, in which the default paths
+where the log files are found should be set:
+
+[source,yaml]
+----
+var:
+  - name: paths
+    default:
+      - /example/test.log*
+    os.darwin:
+      - /usr/local/example/test.log*
+      - /example/test.log*
+    os.windows:
+      - c:/programdata/example/logs/test.log*
+----
+
+There's quite a lot going on in the above, so let's break it down:
+
+* The name of the variable is `paths` and the default value is
+  `["/example/test.log*"]`.
```

Comment:
It could be confusing that here the compact array form is used, but above it the example not. For people familiar with Yaml, it is obvious.